File: _old_csharp_backup/LeonEdBackend/Services/Implementations/DashboardService.cs
using Microsoft.EntityFrameworkCore;
using LeonEdBackend.Data;
using LeonEdBackend.DTOs.Common;
using LeonEdBackend.DTOs.DashboardDtos;
using LeonEdBackend.DTOs.TeacherDtos;
using LeonEdBackend.Helpers;
using LeonEdBackend.Services.Interfaces;

namespace LeonEdBackend.Services.Implementations
{
    public class DashboardService : IDashboardService
    {
        private readonly AppDbContext _context;
        public DashboardService(AppDbContext context) { _context = context; }

        public async Task<ApiResponse<SchoolDashboardResponse>> GetSchoolDashboard(Guid schoolId)
        {
            var school = await _context.Schools.FirstOrDefaultAsync(s => s.Id == schoolId);
            if (school == null) return ApiResponse<SchoolDashboardResponse>.FailResponse("School not found.");

            var currentSession = await _context.AcademicSessions.FirstOrDefaultAsync(s => s.SchoolId == schoolId && s.IsCurrent);
            var currentTerm = currentSession != null ? await _context.Terms.FirstOrDefaultAsync(t => t.AcademicSessionId == currentSession.Id && t.IsCurrent) : null;

            var response = new SchoolDashboardResponse
            {
                SchoolId = school.Id,
                SchoolName = school.Name,
                SubscriptionPlan = school.SubscriptionPlan.ToString(),
                TotalStudents = await _context.Students.CountAsync(s => s.SchoolId == schoolId && s.Status == StudentStatus.Active),
                TotalTeachers = await _context.Teachers.CountAsync(t => t.SchoolId == schoolId && t.IsActive),
                TotalClasses = await _context.Classes.CountAsync(c => c.SchoolId == schoolId),
                TotalSubjects = await _context.Subjects.CountAsync(s => s.SchoolId == schoolId),
                MaxStudents = school.MaxStudents,
                MaxTeachers = school.MaxTeachers,
                CurrentSession = currentSession?.Name,
                CurrentTerm = currentTerm?.TermNumber.ToString()
            };
            return ApiResponse<SchoolDashboardResponse>.SuccessResponse(response);
        }

        public async Task<ApiResponse<SuperAdminDashboardResponse>> GetSuperAdminDashboard()
        {
            var response = new SuperAdminDashboardResponse
            {
                TotalSchools = await _context.Schools.CountAsync(),
                ActiveSchools = await _context.Schools.CountAsync(s => s.IsActive),
                SuspendedSchools = await _context.Schools.CountAsync(s => !s.IsActive),
                TotalStudentsAcrossSchools = await _context.Students.CountAsync(s => s.Status == StudentStatus.Active),
                TotalTeachersAcrossSchools = await _context.Teachers.CountAsync(t => t.IsActive),
                PlanBreakdown = new PlanBreakdown
                {
                    FreeSchools = await _context.Schools.CountAsync(s => s.SubscriptionPlan == SubscriptionPlan.Free),
                    PlusSchools = await _context.Schools.CountAsync(s => s.SubscriptionPlan == SubscriptionPlan.Plus),
                    PremiumSchools = await _context.Schools.CountAsync(s => s.SubscriptionPlan == SubscriptionPlan.Premium)
                }
            };
            return ApiResponse<SuperAdminDashboardResponse>.SuccessResponse(response);
        }

        public async Task<ApiResponse<TeacherDashboardResponse>> GetTeacherDashboard(Guid schoolId, Guid userId)
        {
            var teacher = await _context.Teachers
                .Include(t => t.SubjectAssignments)
                    .ThenInclude(a => a.Subject)
                .Include(t => t.SubjectAssignments)
                    .ThenInclude(a => a.Class)
                .FirstOrDefaultAsync(t => t.UserId == userId && t.SchoolId == schoolId);

            if (teacher == null) return ApiResponse<TeacherDashboardResponse>.FailResponse("Teacher profile not found.");

            var currentSession = await _context.AcademicSessions.FirstOrDefaultAsync(s => s.SchoolId == schoolId && s.IsCurrent);
            var currentTerm = currentSession != null ? await _context.Terms.FirstOrDefaultAsync(t => t.AcademicSessionId == currentSession.Id && t.IsCurrent) : null;

            var assignments = teacher.SubjectAssignments.Select(a => new TeacherAssignmentResponse
            {
                Id = a.Id,
                SubjectId = a.SubjectId,
                SubjectName = a.Subject?.Name ?? "",
                ClassId = a.ClassId,
                ClassName = a.Class != null ? $"{a.Class.Name} {a.Class.Arm}".Trim() : ""
            }).ToList();

            var response = new TeacherDashboardResponse
            {
                TeacherId = teacher.Id,
                FullName = teacher.FullName,
                TotalAssignedSubjects = assignments.Select(a => a.SubjectId).Distinct().Count(),
                TotalAssignedClasses = assignments.Select(a => a.ClassId).Distinct().Count(),
                CurrentSession = currentSession?.Name,
                CurrentTerm = currentTerm?.TermNumber.ToString(),
                Assignments = assignments
            };

            return ApiResponse<TeacherDashboardResponse>.SuccessResponse(response);
        }

        public async Task<ApiResponse<StudentDashboardResponse>> GetStudentDashboard(Guid schoolId, Guid userId)
        {
            var student = await _context.Students
                .Include(s => s.Class)
                .FirstOrDefaultAsync(s => s.UserId == userId && s.SchoolId == schoolId);

            if (student == null) return ApiResponse<StudentDashboardResponse>.FailResponse("Student profile not found.");

            var currentSession = await _context.AcademicSessions.FirstOrDefaultAsync(s => s.SchoolId == schoolId && s.IsCurrent);
            var currentTerm = currentSession != null ? await _context.Terms.FirstOrDefaultAsync(t => t.AcademicSessionId == currentSession.Id && t.IsCurrent) : null;

            var response = new StudentDashboardResponse
            {
                StudentId = student.Id,
                FullName = student.FullName,
                AdmissionNumber = student.AdmissionNumber,
                ClassName = student.Class?.Name,
                ClassArm = student.Class?.Arm,
                CurrentSession = currentSession?.Name,
                CurrentTerm = currentTerm?.TermNumber.ToString()
            };

            return ApiResponse<StudentDashboardResponse>.SuccessResponse(response);
        }
    }
}

File: src/services/dashboard.service.ts
import { prisma } from "../config/db";
import { successResponse, failResponse } from "../utils/response";

export class DashboardService {

  static async getSchoolDashboard(schoolId: string) {
    const school = await prisma.school.findUnique({
      where: { id: schoolId },
      include: { plan: true },
    });

    if (!school) {
      return failResponse("School not found.");
    }

    const currentSession = await prisma.academicSession.findFirst({
      where: { schoolId, isCurrent: true },
    });

    const currentTerm = currentSession
      ? await prisma.term.findFirst({
          where: { academicSessionId: currentSession.id, isCurrent: true },
        })
      : null;

    const totalStudents = await prisma.student.count({
      where: { schoolId, status: "Active" },
    });

    const totalTeachers = await prisma.teacher.count({
      where: { schoolId, isActive: true },
    });

    const totalClasses = await prisma.class.count({
      where: { schoolId },
    });

    const totalSubjects = await prisma.subject.count({
      where: { schoolId },
    });

    const pendingResults = await prisma.result.count({
      where: {
        schoolId,
        status: { in: ["Draft", "Submitted"] },
        ...(currentTerm ? { termId: currentTerm.id } : {}),
      },
    });

    let termProgress = 0;
    if (currentTerm) {
      const start = currentTerm.startDate.getTime();
      const end = currentTerm.endDate.getTime();
      const now = Date.now();
      if (end > start) {
        const progress = ((now - start) / (end - start)) * 100;
        termProgress = Math.max(0, Math.min(100, Math.round(progress)));
      }
    }

    return successResponse({
      schoolId: school.id,
      schoolName: school.name,
      subscriptionPlan: school.plan?.name || "Free",
      totalStudents,
      totalTeachers,
      totalFaculty: totalTeachers,
      totalClasses,
      totalSubjects,
      maxStudents: school.plan?.maxStudents ?? 100,
      maxTeachers: school.plan?.maxTeachers ?? 20,
      currentSession: currentSession?.name || null,
      currentTerm: currentTerm?.termNumber || null,
      pendingResults,
      termProgress,
    });
  }

  static async getSuperAdminDashboard() {
    const totalSchools = await prisma.school.count();
    const activeSchools = await prisma.school.count({ where: { isActive: true } });
    const suspendedSchools = await prisma.school.count({ where: { isActive: false } });

    const totalStudentsAcrossSchools = await prisma.student.count({
      where: { status: "Active" },
    });

    const totalTeachersAcrossSchools = await prisma.teacher.count({
      where: { isActive: true },
    });

    const plans = await prisma.paymentPlan.findMany({
      include: {
        _count: {
          select: { schools: true },
        },
      },
    });

    const freeSchoolsCount = await prisma.school.count({
      where: { planId: null },
    });

    const planBreakdown: Record<string, number> = {};
    planBreakdown["Free"] = freeSchoolsCount;
    for (const p of plans) {
      planBreakdown[p.name] = p._count.schools;
    }

    return successResponse({
      totalSchools,
      activeSchools,
      suspendedSchools,
      totalStudentsAcrossSchools,
      totalTeachersAcrossSchools,
      planBreakdown,
    });
  }

  static async getTeacherDashboard(schoolId: string, userId: string) {
    const teacher = await prisma.teacher.findFirst({
      where: { userId, schoolId },
      include: {
        subjectAssignments: {
          include: {
            subject: true,
            class: {
              include: {
                students: {
                  where: { status: "Active" },
                  select: { id: true },
                },
              },
            },
          },
        },
      },
    });

    if (!teacher) {
      return failResponse("Teacher profile not found.");
    }

    const currentSession = await prisma.academicSession.findFirst({
      where: { schoolId, isCurrent: true },
    });

    const currentTerm = currentSession
      ? await prisma.term.findFirst({
          where: { academicSessionId: currentSession.id, isCurrent: true },
        })
      : null;

    const assignments = teacher.subjectAssignments.map((a) => ({
      id: a.id,
      subjectId: a.subjectId,
      subjectName: a.subject ? a.subject.name : "",
      classId: a.classId,
      className: a.class ? `${a.class.name} ${a.class.arm}`.trim() : "",
      studentCount: a.class?.students?.length || 0,
    }));

    const distinctSubjectIds = new Set(assignments.map((a) => a.subjectId));
    const distinctClassIds = new Set(assignments.map((a) => a.classId));

    // Total unique students across all assigned classes
    const totalStudents = await prisma.student.count({
      where: {
        schoolId,
        status: "Active",
        classId: { in: Array.from(distinctClassIds) },
      },
    });

    return successResponse({
      teacherId: teacher.id,
      fullName: teacher.fullName,
      profilePictureUrl: teacher.profilePictureUrl || "",
      totalAssignedSubjects: distinctSubjectIds.size,
      totalAssignedClasses: distinctClassIds.size,
      totalStudents,
      currentSession: currentSession?.name || null,
      currentTerm: currentTerm?.termNumber || null,
      assignments,
    });
  }

  static async getStudentDashboard(schoolId: string, userId: string) {
    const student = await prisma.student.findFirst({
      where: { userId, schoolId },
      include: { class: true },
    });

    if (!student) {
      return failResponse("Student profile not found.");
    }

    const currentSession = await prisma.academicSession.findFirst({
      where: { schoolId, isCurrent: true },
    });

    const currentTerm = currentSession
      ? await prisma.term.findFirst({
          where: { academicSessionId: currentSession.id, isCurrent: true },
        })
      : null;

    return successResponse({
      studentId: student.id,
      fullName: student.fullName,
      admissionNumber: student.admissionNumber,
      className: student.class ? student.class.name : null,
      classArm: student.class ? student.class.arm : null,
      currentSession: currentSession?.name || null,
      currentTerm: currentTerm?.termNumber || null,
    });
  }
}
